zonemaster/zonemaster-backend
# Add metrics

## Purpose

Add application level metrics to enable better observability.

## Changes

Use Net::Statsd to send metrics to a statsd compatible receiver. I chose a pushed based metrics system over a pull based metrics collection (e.g. Prometheus / Opentelemetry) because both the API and test agent are forking, making it hard to aggregate the metrics.

The change is purely optional and it is not required to installed any dependencies for users that don't need this feature.

NOTE: This PR is still a proof of concept, I wanted some insight before going further with it.

## How to test this PR

* Install Net::Statsd `cpanm Net::Statsd` 
* Set / export the variables `ZM_STATS_HOST` and `ZM_STATS_PORT` to the statsd receiver host and port (upd). The port can be omitted in which case it will default to 8125.
* Start the API and Test Agent.

Using the [statsd exporter](https://github.com/prometheus/statsd_exporter/releases) as receiver with the following mapping:
```yaml
mappings:
- match: zonemaster.rpcapi.requests.*.*
  name: zonemaster_rpcapi_requests_total
  labels:
    method: $1
    status: $2
```
we should see some metrics starting with `zonemaster_*`, e.g.:
```
# HELP zonemaster_rpcapi_requests_total Metric autogenerated by statsd_exporter.
# TYPE zonemaster_rpcapi_requests_total counter
zonemaster_rpcapi_requests_total{method="get_test_results",status="RPC_INTERNAL_ERROR"} 1
zonemaster_rpcapi_requests_total{method="get_test_results",status="RPC_SUCCESS"} 1
zonemaster_rpcapi_requests_total{method="start_domain_test",status="RPC_SUCCESS"} 3
zonemaster_rpcapi_requests_total{method="test_progress",status="RPC_SUCCESS"} 63
# HELP zonemaster_testagent_maximum_processes Metric autogenerated by statsd_exporter.
# TYPE zonemaster_testagent_maximum_processes gauge
zonemaster_testagent_maximum_processes 40
# HELP zonemaster_testagent_running_processes Metric autogenerated by statsd_exporter.
# TYPE zonemaster_testagent_running_processes gauge
zonemaster_testagent_running_processes 1
# HELP zonemaster_testagent_tests_completed Metric autogenerated by statsd_exporter.
# TYPE zonemaster_testagent_tests_completed counter
zonemaster_testagent_tests_completed 1
# HELP zonemaster_testagent_tests_died Metric autogenerated by statsd_exporter.
# TYPE zonemaster_testagent_tests_died counter
zonemaster_testagent_tests_died 1
# HELP zonemaster_testagent_tests_started Metric autogenerated by statsd_exporter.
# TYPE zonemaster_testagent_tests_started counter
zonemaster_testagent_tests_started 3
```
